std/permutation/poseidon2/gkr.go

```diff
@@ -0,0 +1,400 @@
+// they must be padded to a power of 2
+func defineCircuit(insLeft, insRight []frontend.Variable) (*gkr.API, constraint.GkrVariable, error) {
+	// variable indexes
+	const (
+		xI = iota
+		yI
+	)
+
+	// poseidon2 parameters
+	roundKeysFr := poseidon2Bls12377.GetDefaultParameters().RoundKeys
+	params := poseidon2Bls12377.GetDefaultParameters().String()
+	rF := poseidon2Bls12377.GetDefaultParameters().NbFullRounds
+	rP := poseidon2Bls12377.GetDefaultParameters().NbPartialRounds
+	halfRf := rF / 2
+
+	gkrApi := gkr.NewApi()
+
+	x, err := gkrApi.Import(insLeft)
+	if err != nil {
+		return nil, -1, err
+	}
+	y, err := gkrApi.Import(insRight)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	// unique names for linear rounds
+	gateNameLinear := func(varI, round int) string {
+		return fmt.Sprintf("x%d-l-op-round=%d;%s", varI, round, params)
+	}
+
+	// the s-Box gates: u¹⁷ = (u⁴)⁴ * u
+	gkr.Gates["pow4"] = pow4Gate{}
+	gkr.Gates["pow4Times"] = pow4TimesGate{}
+
+	// *** helper functions to register and apply gates ***
+
+	// Poseidon2 is a sequence of additions, exponentiations (s-Box), and linear operations
+	// but here we group the operations so that every round consists of a degree-1 operation followed by the s-Box
+	// this allows for more efficient result sharing among the gates
+	// but also breaks the uniformity of the circuit a bit, in that the matrix operation
+	// in every round comes from the previous (canonical) round.
+
+	// apply the s-Box to u
+	sBox := func(u constraint.GkrVariable) constraint.GkrVariable {
+		v := gkrApi.NamedGate("pow4", u)           // u⁴
+		return gkrApi.NamedGate("pow4Times", v, u) // u¹⁷
+	}
+
+	// register and apply external matrix multiplication and round key addition
+	// round dependent due to the round key
+	extKeySBox := func(round, varI int, a, b constraint.GkrVariable) constraint.GkrVariable {
+		gate := gateNameLinear(varI, round)
+		gkr.Gates[gate] = &extKeyGate{
+			roundKey: frToInt(&roundKeysFr[round][varI]),
+		}
+		return sBox(gkrApi.NamedGate(gate, a, b))
+	}
+
+	// register and apply external matrix multiplication and round key addition
+	// then apply the s-Box
+	// for the second variable
+	// round independent due to the round key
+	intKeySBox2 := func(round int, a, b constraint.GkrVariable) constraint.GkrVariable {
+		gate := gateNameLinear(yI, round)
+		gkr.Gates[gate] = &intKeyGate2{
+			roundKey: frToInt(&roundKeysFr[round][1]),
+		}
+		return sBox(gkrApi.NamedGate(gate, a, b))
+	}
+
+	// apply a full round
+	fullRound := func(i int) {
+		x1 := extKeySBox(i, xI, x, y)      // TODO inline this
+		x, y = x1, extKeySBox(i, yI, y, x) // the external matrix is symmetric so we can use the same gate with inputs swapped
+	}
+
+	// *** construct the circuit ***
+
+	for i := range halfRf {
+		fullRound(i)
+	}
+
+	{
+		// i = halfRf: first partial round
+		// still using the external matrix, since the linear operation still belongs to a full (canonical) round
+		x1 := extKeySBox(halfRf, xI, x, y)
```

Comment:
I don't fully understand why here we do `extKeySBox` not `extGate2`? Imo we have done `rf/2` times `extKeySbox`, so we should be only missing `ext` for full input?

Reply:
Oh, I think I understand now, it is essentially the first loop of
```go
	for i := rf; i < rf+h.params.nbPartialRounds; i++ {
		// one round = matMulInternal(sBox_sparse(addRoundKey))
		h.addRoundKeyInPlace(i, input)
		h.sBox(0, input)
		h.matMulInternalInPlace(input)
	}
```
without the `h.matMulInternalInPlace`.